### PR TITLE
MosekSolver result get dual solution for bounding box constraints.

### PR DESCRIPTION
**Pull request by hongkai-dai**
As a step to #13285

This gets the dual solution for only bounding box constraints. I will later send PRs for linear constraints and nonlinear conic constraints.

cc @TobiaMarcucci

---
This change is [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/14303)

### CODE REVIEW

**Review by hongkai-dai — COMMENTED** [review]
+@avalenzu for feature review please, thanks!

> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/14303#-:-MLf7dOg5hJxGGcXiD8X:b-uss1yj)* status: LGTM missing from assignee avalenzu, needs platform reviewer assigned, needs at least two assigned reviewers (waiting on @avalenzu)

**Review by avalenzu — COMMENTED** [review]
Review pass done.

> Reviewed 6 of 6 files at r1.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/14303#-:-MLmfLin0Z9mXpGiMMmX:b-l0f5b4)* status: 15 unresolved discussions, LGTM missing from assignee avalenzu, needs platform reviewer assigned, needs at least two assigned reviewers (waiting on @hongkai-dai)

___
*[solvers/mosek_solver.cc, line 95 at r1](https://reviewable.io/reviews/robotlocomotion/drake/14303#-MLmO32z6j0iCVYGchqm:-MLmO32z6j0iCVYGchqn:bxwh6ru) ([raw file](https://github.com/robotlocomotion/drake/blob/af5450b96c4ec80db0e69bfbc1d5dc54d648452d/solvers/mosek_solver.cc#L95)):*
> ```cpp
> // for more information.
> enum class DualVarType {
>   kSlc,  ///< linear constraint lower bound.
> ```

It might be easier on the reader if we switched the names and comments here:
```
enum class DualVarType {
  kLinearConstraintLowerBound,  ///< Corresponds to Mosek's slc.
  kLinearConstraintUpperBound,  ///< Corresponds to Mosek's suc.
  kVariableLowerBound,          ///< Corresponds to Mosek's slx.
  kVariableUpperBound,          ///< Corresponds to Mosek's sux.
  kNonlinearConic,              ///< Corresponds to Mosek's snx.
};
```
That way we restrict the cryptic names to the `enum` declaration.
___
*[solvers/mosek_solver.cc, line 453 at r1](https://reviewable.io/reviews/robotlocomotion/drake/14303#-MLmQJs95USG547YDFS-:-MLmQJs95USG547YDFS0:bgr3imi) ([raw file](https://github.com/robotlocomotion/drake/blob/af5450b96c4ec80db0e69bfbc1d5dc54d648452d/solvers/mosek_solver.cc#L453)):*
> ```cpp
>                        std::pair<std::vector<ConstraintDualIndex>,
>                                  std::vector<ConstraintDualIndex>>>*
>         bb_con_dual_indices,
> ```

Given the scope of this variable, I think you could get rid of abbreviations *and* make the name shorter by simply calling it `dual_indices`.x
___
*[solvers/mosek_solver.cc, line 543 at r1](https://reviewable.io/reviews/robotlocomotion/drake/14303#-MLmVSz10Kmh0n5ctnwh:-MLmVSz10Kmh0n5ctnwi:b-hvydgs) ([raw file](https://github.com/robotlocomotion/drake/blob/af5450b96c4ec80db0e69bfbc1d5dc54d648452d/solvers/mosek_solver.cc#L543)):*
> ```cpp
>   // Make sure after calling AddLinearConstraintToMosek, the number of newly
>   // added linear constraints is bounded_matrix_var_count. This is important
>   // since we determine the index of the dual variable for bounds on matrix
> ```

nit: s/since/because/
___
*[solvers/mosek_solver.cc, line 544 at r1](https://reviewable.io/reviews/robotlocomotion/drake/14303#-MLmSL-y7yJ1SZsNYWR2:-MLmSL-y7yJ1SZsNYWR3:b-c0abok) ([raw file](https://github.com/robotlocomotion/drake/blob/af5450b96c4ec80db0e69bfbc1d5dc54d648452d/solvers/mosek_solver.cc#L544)):*
> ```cpp
>   // added linear constraints is bounded_matrix_var_count. This is important
>   // since we determine the index of the dual variable for bounds on matrix
>   // variables, based on the order of adding linear constraints in
> ```

nit: remove comma after "variables"
___
*[solvers/mosek_solver.cc, line 578 at r1](https://reviewable.io/reviews/robotlocomotion/drake/14303#-MLmYQI8ASqjvKyAk1sp:-MLmYQI979WvgDCPc77i:b-r2yadk) ([raw file](https://github.com/robotlocomotion/drake/blob/af5450b96c4ec80db0e69bfbc1d5dc54d648452d/solvers/mosek_solver.cc#L578)):*
> ```cpp
>       auto it1 =
>           decision_variable_index_to_mosek_nonmatrix_variable.find(var_index);
>       if (it1 != decision_variable_index_to_mosek_nonmatrix_variable.end()) {
> ```

I think you should be able reduce duplication by factoring out the active/inactive checks as shown below
```c++
int dual_variable_index_if_active = -1;                                         
if (it1 != decision_variable_index_to_mosek_nonmatrix_variable.end()) {         
  // Add the dual variables for the constraint registered as bounds on          
  // Mosek non-matrix variables.                                                
  lower_bound_duals[i].type = DualVarType::kSlx;                                
  upper_bound_duals[i].type = DualVarType::kSux;                                
  dual_variable_index_if_active = it1->second;                                  
} else {                                                                        
  // This variable is a Mosek matrix variable. The bound on this variable       
  // is imposed as a linear constraint.                                         
  DRAKE_DEMAND(decision_variable_index_to_mosek_matrix_variable.count(          
                   var_index) > 0);                                             
  lower_bound_duals[i].type = DualVarType::kSlc;                                
  upper_bound_duals[i].type = DualVarType::kSuc;                                
  const int linear_constraint_index =                                           
      num_linear_constraints_before +                                           
      var_in_bounded_matrix_vars.at(binding.variables()(i).get_id());           
  dual_variable_index_if_active = linear_constraint_index;                      
}                                                                               
if (binding.evaluator()->lower_bound()(i) == x_lb[var_index]) {                 
  // This lower bound can be active.                                            
  lower_bound_duals[i].index = dual_variable_index_if_active;                   
} else {                                                                        
  // This lower bound can't be active, set the index to -1 to indicate          
  // inactiveness.                                                              
  lower_bound_duals[i].index = -1;                                              
}                                                                               
if (binding.evaluator()->upper_bound()(i) == x_ub[var_index]) {                 
  // This upper bound can be active.                                            
  upper_bound_duals[i].index = dual_variable_index_if_active;                   
} else {                                                                        
  // This upper bound can't be active, set the index to -1 to indicate          
  // inactiveness.                                                              
  upper_bound_duals[i].index = -1;                                              
}
```
___
*[solvers/mosek_solver.cc, line 1382 at r1](https://reviewable.io/reviews/robotlocomotion/drake/14303#-MLmbkFZCWNnQX1RthuD:-MLmbkF_EskVCgs65U_q:b61blv5) ([raw file](https://github.com/robotlocomotion/drake/blob/af5450b96c4ec80db0e69bfbc1d5dc54d648452d/solvers/mosek_solver.cc#L1382)):*
> ```cpp
> void SetBoundingBoxDualSolution(
>     const std::vector<Binding<BoundingBoxConstraint>>& constraints,
>     const std::vector<MSKrealt>& slx, const std::vector<MSKrealt>& sux,
> ```

I know these are the Mosek terms, but would it be possible to use more descriptive names for these variables?
___
*[solvers/mosek_solver.cc, line 1393 at r1](https://reviewable.io/reviews/robotlocomotion/drake/14303#-MLmbuXr7zDfDsiq0Nvv:-MLmbuXr7zDfDsiq0Nvw:b-gqfjyj) ([raw file](https://github.com/robotlocomotion/drake/blob/af5450b96c4ec80db0e69bfbc1d5dc54d648452d/solvers/mosek_solver.cc#L1393)):*
> ```cpp
>          const std::vector<MSKrealt>& mosek_dual_upper, double* dual_sol) {
>         if (mosek_dual_index != -1) {
>           // Mosek defines all dual solution as non-negative. But we use
> ```

nit: s/solution/solutions/
___
*[solvers/mosek_solver.cc, line 1393 at r1](https://reviewable.io/reviews/robotlocomotion/drake/14303#-MLmbzba70iGXfcrsZWf:-MLmbzba70iGXfcrsZWg:bt1jbxp) ([raw file](https://github.com/robotlocomotion/drake/blob/af5450b96c4ec80db0e69bfbc1d5dc54d648452d/solvers/mosek_solver.cc#L1393)):*
> ```cpp
>          const std::vector<MSKrealt>& mosek_dual_upper, double* dual_sol) {
>         if (mosek_dual_index != -1) {
>           // Mosek defines all dual solution as non-negative. But we use
> ```

nit: s/But/However,/
___
*[solvers/mosek_solver.cc, line 1394 at r1](https://reviewable.io/reviews/robotlocomotion/drake/14303#-MLmc6fx16-nYtFvYt2t:-MLmc6fx16-nYtFvYt2u:b-seqh5t) ([raw file](https://github.com/robotlocomotion/drake/blob/af5450b96c4ec80db0e69bfbc1d5dc54d648452d/solvers/mosek_solver.cc#L1394)):*
> ```cpp
>         if (mosek_dual_index != -1) {
>           // Mosek defines all dual solution as non-negative. But we use
>           // "reduced cost" as the dual solution, so the dual solution for lower
> ```

nit: s/for lower/for a lower/
___
*[solvers/mosek_solver.cc, line 1395 at r1](https://reviewable.io/reviews/robotlocomotion/drake/14303#-MLmcAJUEaQBLii3iJmv:-MLmcAJV6fTrDs3xsG2h:bqwrdsf) ([raw file](https://github.com/robotlocomotion/drake/blob/af5450b96c4ec80db0e69bfbc1d5dc54d648452d/solvers/mosek_solver.cc#L1395)):*
> ```cpp
>           // Mosek defines all dual solution as non-negative. But we use
>           // "reduced cost" as the dual solution, so the dual solution for lower
>           // bound should be non-negative, while the dual solution for upper
> ```

nit: s/for upper/for an upper/
___
*[solvers/mosek_solver.cc, line 1434 at r1](https://reviewable.io/reviews/robotlocomotion/drake/14303#-MLmceri4wUTYv2wqJyc:-MLmceri4wUTYv2wqJyd:b-x5ibqp) ([raw file](https://github.com/robotlocomotion/drake/blob/af5450b96c4ec80db0e69bfbc1d5dc54d648452d/solvers/mosek_solver.cc#L1434)):*
> ```cpp
> 
> MSKrescodee SetDualSolution(
>     const MSKtask_t& task, MSKsoltypee whichsol,
> ```

nit: per the style guide this should be `which_sol`.
___
*[solvers/mosek_solver.cc, line 1439 at r1](https://reviewable.io/reviews/robotlocomotion/drake/14303#-MLmcog50JxySvlTRo4o:-MLmcog50JxySvlTRo4p:bqxc2m2) ([raw file](https://github.com/robotlocomotion/drake/blob/af5450b96c4ec80db0e69bfbc1d5dc54d648452d/solvers/mosek_solver.cc#L1439)):*
> ```cpp
>                              std::pair<std::vector<ConstraintDualIndex>,
>                                        std::vector<ConstraintDualIndex>>>&
>         bb_con_dual_indices,
> ```

As above, I think you could just use `dual_indices` here.
___
*[solvers/mosek_solver.cc, line 1447 at r1](https://reviewable.io/reviews/robotlocomotion/drake/14303#-MLmcyuV4-GE4ecd4jpr:-MLmcyuV4-GE4ecd4jps:b-3qa3al) ([raw file](https://github.com/robotlocomotion/drake/blob/af5450b96c4ec80db0e69bfbc1d5dc54d648452d/solvers/mosek_solver.cc#L1447)):*
> ```cpp
>     return rescode;
>   }
>   std::vector<MSKrealt> slx(num_mosek_vars);
> ```

Could these be descriptive names too?
___
*[solvers/test/quadratic_program_examples.cc, line 540 at r1](https://reviewable.io/reviews/robotlocomotion/drake/14303#-MLmeTJh5G3DjHc_k04V:-MLmeTJh5G3DjHc_k04W:boaamyu) ([raw file](https://github.com/robotlocomotion/drake/blob/af5450b96c4ec80db0e69bfbc1d5dc54d648452d/solvers/test/quadratic_program_examples.cc#L540)):*
> ```cpp
> 
> void TestQPDualSolution3(const SolverInterface& solver, double tol1,
>                          double tol2) {
> ```

Why does this need two different tolerances? It looks like maybe one is for checks on variable values and the other is for checks on costs? If so, could they be renamed accordingly?
___
*[solvers/test/semidefinite_program_examples.h, line 112 at r1](https://reviewable.io/reviews/robotlocomotion/drake/14303#-MLmeyW726ZQMJAubdTJ:-MLmeyW726ZQMJAubdTK:b6c7kus) ([raw file](https://github.com/robotlocomotion/drake/blob/af5450b96c4ec80db0e69bfbc1d5dc54d648452d/solvers/test/semidefinite_program_examples.h#L112)):*
> ```objc
> 
> /**
>  * Test a simple PSD
> ```

nit: s/PSD/SDP/

**Review by hongkai-dai — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/14303)* status: 7 unresolved discussions, LGTM missing from assignee avalenzu, needs platform reviewer assigned, needs at least two assigned reviewers (waiting on @avalenzu)

___
*[solvers/mosek_solver.cc, line 95 at r1](https://reviewable.io/reviews/robotlocomotion/drake/14303#-MLmO32z6j0iCVYGchqm:-MLnUKKV9qe2tVj9q7rV:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/af5450b96c4ec80db0e69bfbc1d5dc54d648452d/solvers/mosek_solver.cc#L95)):*
<details><summary><i>Previously, avalenzu (Andres Valenzuela) wrote…</i></summary><blockquote>

It might be easier on the reader if we switched the names and comments here:
```
enum class DualVarType {
  kLinearConstraintLowerBound,  ///< Corresponds to Mosek's slc.
  kLinearConstraintUpperBound,  ///< Corresponds to Mosek's suc.
  kVariableLowerBound,          ///< Corresponds to Mosek's slx.
  kVariableUpperBound,          ///< Corresponds to Mosek's sux.
  kNonlinearConic,              ///< Corresponds to Mosek's snx.
};
```
That way we restrict the cryptic names to the `enum` declaration.
</blockquote></details>

Done.
___
*[solvers/mosek_solver.cc, line 453 at r1](https://reviewable.io/reviews/robotlocomotion/drake/14303#-MLmQJs95USG547YDFS-:-MLnUYRGAA2sLi9lUghh:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/af5450b96c4ec80db0e69bfbc1d5dc54d648452d/solvers/mosek_solver.cc#L453)):*
<details><summary><i>Previously, avalenzu (Andres Valenzuela) wrote…</i></summary><blockquote>

Given the scope of this variable, I think you could get rid of abbreviations *and* make the name shorter by simply calling it `dual_indices`.x
</blockquote></details>

Done.
___
*[solvers/mosek_solver.cc, line 578 at r1](https://reviewable.io/reviews/robotlocomotion/drake/14303#-MLmYQI8ASqjvKyAk1sp:-MLnVili6vs3-Ljj_gqm:boyexb3) ([raw file](https://github.com/robotlocomotion/drake/blob/af5450b96c4ec80db0e69bfbc1d5dc54d648452d/solvers/mosek_solver.cc#L578)):*
<details><summary><i>Previously, avalenzu (Andres Valenzuela) wrote…</i></summary><blockquote>

I think you should be able reduce duplication by factoring out the active/inactive checks as shown below
```c++
int dual_variable_index_if_active = -1;                                         
if (it1 != decision_variable_index_to_mosek_nonmatrix_variable.end()) {         
  // Add the dual variables for the constraint registered as bounds on          
  // Mosek non-matrix variables.                                                
  lower_bound_duals[i].type = DualVarType::kSlx;                                
  upper_bound_duals[i].type = DualVarType::kSux;                                
  dual_variable_index_if_active = it1->second;                                  
} else {                                                                        
  // This variable is a Mosek matrix variable. The bound on this variable       
  // is imposed as a linear constraint.                                         
  DRAKE_DEMAND(decision_variable_index_to_mosek_matrix_variable.count(          
                   var_index) > 0);                                             
  lower_bound_duals[i].type = DualVarType::kSlc;                                
  upper_bound_duals[i].type = DualVarType::kSuc;                                
  const int linear_constraint_index =                                           
      num_linear_constraints_before +                                           
      var_in_bounded_matrix_vars.at(binding.variables()(i).get_id());           
  dual_variable_index_if_active = linear_constraint_index;                      
}                                                                               
if (binding.evaluator()->lower_bound()(i) == x_lb[var_index]) {                 
  // This lower bound can be active.                                            
  lower_bound_duals[i].index = dual_variable_index_if_active;                   
} else {                                                                        
  // This lower bound can't be active, set the index to -1 to indicate          
  // inactiveness.                                                              
  lower_bound_duals[i].index = -1;                                              
}                                                                               
if (binding.evaluator()->upper_bound()(i) == x_ub[var_index]) {                 
  // This upper bound can be active.                                            
  upper_bound_duals[i].index = dual_variable_index_if_active;                   
} else {                                                                        
  // This upper bound can't be active, set the index to -1 to indicate          
  // inactiveness.                                                              
  upper_bound_duals[i].index = -1;                                              
}
```
</blockquote></details>

Done. Thanks!
___
*[solvers/mosek_solver.cc, line 1382 at r1](https://reviewable.io/reviews/robotlocomotion/drake/14303#-MLmbkFZCWNnQX1RthuD:-MLnWZeuEp32qJAM3Z8G:baxftia) ([raw file](https://github.com/robotlocomotion/drake/blob/af5450b96c4ec80db0e69bfbc1d5dc54d648452d/solvers/mosek_solver.cc#L1382)):*
<details><summary><i>Previously, avalenzu (Andres Valenzuela) wrote…</i></summary><blockquote>

I know these are the Mosek terms, but would it be possible to use more descriptive names for these variables?
</blockquote></details>

I actually prefer to mosek terms here, as it indicates where these variables come from. (I expect any future maintainer of this code to know the Mosek code before making changes to this code). I added some documentations to explain the variables.
___
*[solvers/mosek_solver.cc, line 1439 at r1](https://reviewable.io/reviews/robotlocomotion/drake/14303#-MLmcog50JxySvlTRo4o:-MLnXI1E-MbvEARt88lw:bavuhcm) ([raw file](https://github.com/robotlocomotion/drake/blob/af5450b96c4ec80db0e69bfbc1d5dc54d648452d/solvers/mosek_solver.cc#L1439)):*
<details><summary><i>Previously, avalenzu (Andres Valenzuela) wrote…</i></summary><blockquote>

As above, I think you could just use `dual_indices` here.
</blockquote></details>

This `SetDualSolution` function will accept more constraints, like linear constraints, psd cone constraint, etc. So it will have input variables like `linear_con_dual_indices`, `psd_con_dual_indices` in the future. I added a TODO here.
___
*[solvers/mosek_solver.cc, line 1447 at r1](https://reviewable.io/reviews/robotlocomotion/drake/14303#-MLmcyuV4-GE4ecd4jpr:-MLnY9XACcroAxGOpOVw:b-m3dntu) ([raw file](https://github.com/robotlocomotion/drake/blob/af5450b96c4ec80db0e69bfbc1d5dc54d648452d/solvers/mosek_solver.cc#L1447)):*
<details><summary><i>Previously, avalenzu (Andres Valenzuela) wrote…</i></summary><blockquote>

Could these be descriptive names too?
</blockquote></details>

I actually prefer to the mosek variable name, to indicate where these variables come from. I added more documentation.
___
*[solvers/test/quadratic_program_examples.cc, line 540 at r1](https://reviewable.io/reviews/robotlocomotion/drake/14303#-MLmeTJh5G3DjHc_k04V:-MLnZKXd6opjPgnzf5Rr:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/af5450b96c4ec80db0e69bfbc1d5dc54d648452d/solvers/test/quadratic_program_examples.cc#L540)):*
<details><summary><i>Previously, avalenzu (Andres Valenzuela) wrote…</i></summary><blockquote>

Why does this need two different tolerances? It looks like maybe one is for checks on variable values and the other is for checks on costs? If so, could they be renamed accordingly?
</blockquote></details>

Done.

**Review by avalenzu — COMMENTED** [review]
:lgtm: +@soonho-tri for platform review.

> Reviewed 4 of 4 files at r2.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/14303#-:-MLt-rfJ0bF4fQ-nOlHw:b-sf1n7o)* status: LGTM missing from assignee soonho-tri(platform) (waiting on @soonho-tri)

**Review by soonho-tri — COMMENTED** [review]
:lgtm:

> Reviewed 2 of 6 files at r1, 4 of 4 files at r2.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/14303#-:-MLtC9xO3JX-5L1f3RwB:bnfp4nl)* status: 9 unresolved discussions, LGTM missing from assignee avalenzu (waiting on @avalenzu and @hongkai-dai)

___
*[solvers/mosek_solver.cc, line 92 at r2](https://reviewable.io/reviews/robotlocomotion/drake/14303#-MLtBWftBdc3-9eDX800:-MLtBWftBdc3-9eDX801:b-j0w1f0) ([raw file](https://github.com/robotlocomotion/drake/blob/b5e0ed3d9f25eb9e0218b8708512227040c93aa7/solvers/mosek_solver.cc#L92)):*
> ```
> https://docs.mosek.com/9.2/capi/alphabetic-functionalities.html#mosek.task.getsolution
> ```

META, we support mosek-9.0 for now, so let's use the links starting with `https//docs.mosek.com/9.0/*` in this file. We can update them when we use a newer version of mosek.
___
*[solvers/mosek_solver.cc, line 107 at r2](https://reviewable.io/reviews/robotlocomotion/drake/14303#-MLt0sTWEQWZsyVTNmYa:-MLt0sTXDBXNaiSEr0x7:b6vufvj) ([raw file](https://github.com/robotlocomotion/drake/blob/b5e0ed3d9f25eb9e0218b8708512227040c93aa7/solvers/mosek_solver.cc#L107)):*
> ```cpp
>   // Index of the dual variable. We will use -1 to indicate that a constraint
>   // can never be active.
>   int index;
> ```

nit, => `int index{};`
___
*[solvers/mosek_solver.cc, line 612 at r2](https://reviewable.io/reviews/robotlocomotion/drake/14303#-MLt3WqK1uXfHCqDutR6:-MLt3WqK1uXfHCqDutR7:b-m810m5) ([raw file](https://github.com/robotlocomotion/drake/blob/b5e0ed3d9f25eb9e0218b8708512227040c93aa7/solvers/mosek_solver.cc#L612)):*
> ```cpp
>       }
>     }
>     dual_indices->emplace(binding,
> ```

BTW, using c++-17, we can shorten this to 
```c++
dual_indices->try_emplace(binding, lower_bound_duals, upper_bound_duals);
```

**Review by hongkai-dai — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/14303)* status: 8 unresolved discussions, LGTM missing from assignee avalenzu (waiting on @avalenzu and @soonho-tri)

___
*[solvers/mosek_solver.cc, line 92 at r2](https://reviewable.io/reviews/robotlocomotion/drake/14303#-MLtBWftBdc3-9eDX800:-MLthdFi6Cw-LMmGr5WQ:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/b5e0ed3d9f25eb9e0218b8708512227040c93aa7/solvers/mosek_solver.cc#L92)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

> ```
> https://docs.mosek.com/9.2/capi/alphabetic-functionalities.html#mosek.task.getsolution
> ```

META, we support mosek-9.0 for now, so let's use the links starting with `https//docs.mosek.com/9.0/*` in this file. We can update them when we use a newer version of mosek.
</blockquote></details>

Done.
___
*[solvers/mosek_solver.cc, line 612 at r2](https://reviewable.io/reviews/robotlocomotion/drake/14303#-MLt3WqK1uXfHCqDutR6:-MLthmbN3dfRaFJMHzw0:b2cju2i) ([raw file](https://github.com/robotlocomotion/drake/blob/b5e0ed3d9f25eb9e0218b8708512227040c93aa7/solvers/mosek_solver.cc#L612)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

BTW, using c++-17, we can shorten this to 
```c++
dual_indices->try_emplace(binding, lower_bound_duals, upper_bound_duals);
```
</blockquote></details>

Wow, this is fancy!

**Review by soonho-tri — COMMENTED** [review]
> Reviewed 1 of 1 files at r3.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/14303)* status: 7 unresolved discussions, LGTM missing from assignee avalenzu (waiting on @avalenzu)

**Review by avalenzu — COMMENTED** [review]
:lgtm:

> Reviewed 4 of 4 files at r2.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/14303#-:-MLt-rfJ0bF4fQ-nOlHw:bnfp4nl)* status: :shipit: complete! all discussions resolved, LGTM from assignees soonho-tri(platform),avalenzu
